### DISCUSSION TITLE
View() called before WindowSizeMsg

### DISCUSSION THREAD

**Discussion by Robert-M-Muench** [charmbracelet/bubbletea#283] [Q&A] [answered]
> Note that `WindowSizeMsg` is sent before the first render and then again every resize.

_Originally posted by @meowgorithm in https://github.com/charmbracelet/bubbletea/issues/43#issuecomment-758722377_

My current program's `View()` function gets called before I get a `WindowSizeMsg` which makes it impossible to know how to size my output. How can this be possible?

This is the call stack when `View()`is called:
<img width="599" alt="image" src="https://user-images.githubusercontent.com/1826391/162611014-596a2a4b-56d6-4cf0-be99-d1fa7f750a11.png">

And this is the call stack when `Update()` is called:
<img width="596" alt="image" src="https://user-images.githubusercontent.com/1826391/162611204-e28f02c5-8efc-4acd-8813-6271aa256dae.png">

You can see that the `Update()` (line: 547) comes later in the tea.go.StartRetruningModel code as the `View()` (line: 405) call.

**Comment by meowgorithm** [comment, marked as answer]
Hi! The query for the initial window size runs asynchronously so that it doesn't hold up the program startup (since many programs don't need the window size). As a result, it's very possible that the initial view can be called before the window size is queried. If your program does depend on knowing the window size the general pattern is to set a flag on your model when the window size comes in and manage output in your view accordingly.

Here’s a contrived example of how it's commonly done with a notion of application state:

```go
package main

import (
	"fmt"
	"os"

	tea "github.com/charmbracelet/bubbletea"
)

type state int

const (
	initializing state = iota
	ready
)

type model struct {
	state         state
	width, height int
}

func (m model) Init() tea.Cmd {
	return nil
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		return m, tea.Quit
	case tea.WindowSizeMsg:
		m.width, m.height = msg.Width, msg.Height
		m.state = ready
	}
	return m, nil
}

func (m model) View() string {
	if m.state == initializing {
		return "Initializing..."
	}
	return fmt.Sprintf("This is a %dx%d size terminal.", m.width, m.height)
}

func main() {
	if err := tea.NewProgram(model{}).Start(); err != nil {
		fmt.Println("uh oh:", err)
		os.Exit(1)
	}
}
```

Of course, for simple programs you could eliminate the state portion and just check and see if `width` or `height` is greater than `0`.

**Comment by Robert-M-Muench** [comment]
Ok, thanks. The async explains it. Not sure if this is mentioned in the docs somewhere. If not, I think it's worth doing.

**Comment by nghtstr** [comment]
The question I have that goes along with this is if the Update method gets a `tea.WindowSizeMsg` if you never resize the terminal window?

**Reply by meowgorithm** [reply]
When a Bubble Tea program first starts up a `tea.WindowSizeMsg` gets sent to `Update` with the current window dimensions.
